docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+                name: windows-vsphere-template
+      infrastructure:
+        ref:
+          apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+          kind: VSphereClusterTemplate
+          name: vsphere-prod-cluster-template
+   ```
+2. User creates a cluster using the class name and defining the managed topology.
+   ```yaml
+    apiVersion: cluster.x-k8s.io/v1alpha4
+    kind: Cluster
+    metadata:
+      name: foo
+      namespace: bar
+    spec:
+      managed:
+        class: mixed
+        version: v1.19.1
+        controlPlane:
+          replicas: 3
+          labels: {}
+          annotations: {}
+        worker:
+          deployments:
+          - class: linux-worker
+            name: big-pool-of-machines-1
+            replicas: 5
+            labels:
+              # This label is additive to the class' labels,
+              # or if the same label exists, it overwrites it.
+              custom-label: "production"
+          - class: linux-worker
+            name: small-pool-of-machines-1
+            replicas: 1
+          - class: windows-worker
+            name: microsoft-1
+            replicas: 3
+   ```
+3. The Cluster controller checks for the presence of the `spec.managed.class` field. If the field is missing, the Cluster controller behaves in the existing way. Otherwise the controller starts the creation process of the managed topology defined in the steps below.
+4. Cluster and Control plane object creation
+    1. Creates the infrastructure specific/provider cluster using the cluster template referenced in the `ClusterClass.spec.infrastructure.ref` field.
+    1. Add the managed topology label to the provider cluster object:
+       ```yaml
+        cluster.x-k8s.io/managed:<cluster-name>
+       ```
+    1. For the ControlPlane object in `cluster.spec.managed.controlPlane`
+    1. Initializes a control plane object using the control plane template defined in the `ClusterClass.spec.controlPlane.ref field`. Use the name `<cluster-name>`.
+    1. Sets the number of replicas on the control plane object from `spec.managed.controlPlane.replicas`.
+    1. Sets the k8s version on the control plane object from the `spec.managed.version`.
+    1. Add the following labels to the control plane object:
+       ```yaml
+        cluster.x-k8s.io/managed: <cluster-name>
+       ```
+    1. Creates the control plane object.
+    1. Sets the `spec.controlPlaneRef` field for the Cluster object.
+    1. Saves the cluster object in the API server.
+5. Machine deployment object creation
+    1. For each `spec.managed.worker.deployments` item in the list
+    1. Create a name `<cluster-name>-<worker-set-name>` (if too long, hash it)
+    1. Initializes a new MachineDeployment object.
+    1. Sets the `clusterName` field on the MD object
+    1. Sets the `replicas` field on the MD object using `replicas` field for the set of worker nodes.
+    1. Sets the `version` field on the MD object from the `spec.managed.version`.
```

Comment:
We're using different machine templates depending on the k8s version, something that couldn't be done with what's described in this proposal. We can solve it with templating, which is planned for later, so it might not make sense to discuss it here, but I wanted to mention it anyway.

Maybe it would be an option to provide multiple DeploymentClasses and have an (optional) Version field in them and then select one based on version automatically?